### PR TITLE
std.nu: Rewrite assert method

### PR DESCRIPTION
**Pull request by presidento**
It is a common pattern to add message to assert.
Also the error message was not so helpful. (See the difference in the documentation)

### CODE REVIEW

**Comment by presidento** [comment]
@amtoine I think with this change even using `assert` to check equality is more helpful than `assert eq`, see the error messages here:

## With assert

```
assert (1 == 1)
assert (10 == 20)
assert (200 == 200)

Error:
  × Assertion failed: No message
    ╭─[myscript:10:1]
 10 │ assert (1 == 1)
 11 │ assert (10 == 20)
    ·         ────┬───
    ·             ╰── It is not true.
 12 │ assert (200 == 200)
    ╰────

```

## With assert eq

```
assert eq 1 1
assert eq 10 20
assert eq 200 200


Error:
  × left is not equal to right
   ╭─[myscript:5:1]
 5 │     if not $cond {
 6 │         error make {msg: $msg}
   ·         ─────┬────
   ·              ╰── originates from here
 7 │     }
   ╰────
```

I'd consider the value added by `assert eq`.

At first step I did not want to make much change, just send the proposed `assert` command. But if you like it, I can follow up.

**Comment by amtoine** [comment]
@presidento
thanks a lot for this, it's already something that came up in #8368 and i do think having output on the `std assert` familly of commands is a must-have :partying_face: 

i'll have a look after lunch for sure :wink:

File: crates/nu-utils/standard_library/std.nu
Comment (amtoine) [inline crates/nu-utils/standard_library/std.nu:11]: good catch!

File: crates/nu-utils/standard_library/std.nu
Comment (amtoine) [inline crates/nu-utils/standard_library/std.nu:26]: you can use
```rust
$message | default "No message"
```
here :+1: 

i think it's a bit cleaner maybe?

**Review by amtoine — CHANGES_REQUESTED** [review]
loving it :rocket: 

in order to have these beautiful error messages on `assert eq` and `assert ne` as well, could you modify `_assert` instead of `assert`? :yum:

**Comment by presidento** [comment]
Here is the next step. The `_assert` cannot be used, since making errors suppose that it was directly called from user code. (If `assert ne` calls `assert`, it will just show that `$left` and `$right` are not equal.) I extracted the error making step, the only thing which can be common.

I implemented it for `assert eq`, although I think it would be better to use full name, not abbreviation: `assert equal`. For example Nushell uses `str substring`, not `str substr`. Also I think it would better to use `assert not equal` instead of `assert ne`.

I am confused with `assert ne`:

```
>_ 1 != "2"
true

>_ assert ne 1 "2"
Error:
  × left and right operand of `assert eq` should have the same type
   ╭─[myscript.nu:5:1]
 5 │     if not $cond {
 6 │         error make {msg: $msg}
   ·         ─────┬────
   ·              ╰── originates from here
 7 │     }
   ╰────
```

File: crates/nu-utils/standard_library/std.nu
Comment (presidento) [reply crates/nu-utils/standard_library/std.nu]: Good idea, thank you.

Comment (sholderbach) [reply crates/nu-utils/standard_library/std.nu]: I personally dislike the "No message" message. It doesn't really provide guidance and more confusing noise.

Comment (presidento) [reply crates/nu-utils/standard_library/std.nu]: @sholderbach I agree, removed that.

**Comment by amtoine** [comment]
@presidento
no need to rebase and force push, it's normal to apply changes as reviews come to a PR :wink: 

and that's not the most convenient for reviewers to checkout your branch and test what has been done, i think :yum:

**Comment by amtoine** [comment]
i think the names have been chosen to reflect what is done with `rust`, i.e. `assert`, `assert_eq` and `assert_ne` :+1: 
maybe that should be for another PR?

and about `assert ne` it is simply the exact same thing as `assert eq` but with the boolean equality check being reversed, i.e. `==` becomes a `!=` :wink:

File: crates/nu-utils/standard_library/std.nu
Comment (amtoine) [inline crates/nu-utils/standard_library/std.nu:12]: maybe we could only print the `message` when it is given and `Assertion failed.` when it is not?

something like
```diff
-        msg: $"Assertion failed: ($message)",
+        msg: ($message | default "Assertion failed."),
```

Comment (amtoine) [reply crates/nu-utils/standard_library/std.nu:12]: the examples below would have to be modified though

File: crates/nu-utils/standard_library/std.nu
Comment (amtoine) [inline crates/nu-utils/standard_library/std.nu:77]: i'd personnally add a few newlines around the `if` statements :yum:

**Review by amtoine — CHANGES_REQUESTED** [review]

**Comment by presidento** [comment]
In my organization we are allowed to push only perfect commits to master, so we rebase&change commits as long as they are not in the preferred state. If it has no value for this repository, I can leave existing commits as they are and do every change in follow-up commits.

**Comment by amtoine** [comment]
> In my organization we are allowed to push only perfect commits to master, so we rebase&change commits as long as they are not in the preferred state. If it has no value for this repository, I can leave existing commits as they are and do every change in follow-up commits.
>
`nushell` squash all the commits and rebase the commit onto the `main` branch.
the effect is that intermediate commits never end up in the `main` branch, but we can still reopen a PR branch and see a particular commit with a bug introduced for instance :blush:

File: crates/nu-utils/standard_library/std.nu
Comment (presidento) [reply crates/nu-utils/standard_library/std.nu:12]: These are all assertion errors. I think there can be _additional_ information if there is a message. For example:

```
assert equal $code 301 "The http site is permanently redirected to https one"
```

It it fails, it is an assertion error, and the message helps debugging what was the original idea.

File: crates/nu-utils/standard_library/std.nu
Comment (presidento) [reply crates/nu-utils/standard_library/std.nu:77]: I happily follow your personal preferences if they are the official ones here. I assume yes, but I am not sure. (I don't want to be asked for a revert if I follow the guidance.)

Comment (sholderbach) [reply crates/nu-utils/standard_library/std.nu:77]: I see @amtoine is volunteering to write the upcoming `nufmt` to settle all those disputes through technical means :P

File: crates/nu-utils/standard_library/std.nu
Comment (amtoine) [reply crates/nu-utils/standard_library/std.nu:3]: yep!

so i'd say that, when a message is given, there is no need to prepend "Assertion failed:" just print the message :yum: 
but when there is no message given, a simple "Assertion failed." might be enough

in the end, it's the `Assertion failed:` when the message is empty that i find strange :open_mouth:

File: crates/nu-utils/standard_library/std.nu
Comment (amtoine) [reply crates/nu-utils/standard_library/std.nu:67]: on my TODO for some time, coming closer and closer lately :relieved:

Comment (amtoine) [reply crates/nu-utils/standard_library/std.nu:67]: > I happily follow your personal preferences if they are the official ones here. I assume yes, but I am not sure. (I don't want to be asked for a revert if I follow the guidance.)
>
nothing official, but my experience guides me and says it's not good practice to write monolithic blocks of code, maybe except when writing long pipes in shell languages, when the source can be split into several logical pieces :wink:

File: crates/nu-utils/standard_library/std.nu
Comment (amtoine) [inline crates/nu-utils/standard_library/std.nu:90]: you're just missing the type check here :+1: 
just to have `assert eq` and `assert ne` be consistent between each other :yum:

**Review by amtoine — CHANGES_REQUESTED** [review]
one extra thing apart from the type check in `assert ne`, the tests do not pass anymore, so there might be something to investigate there to fix them :relieved: 

```bash
nu crates/nu-utils/standard_library/tests.nu
```
now fails

File: crates/nu-utils/standard_library/std.nu
Comment (presidento) [reply crates/nu-utils/standard_library/std.nu:90]: Sorry, I just forgot to update the tests, it was my mistake.

I also want consistency, see https://github.com/nushell/nushell/pull/8405#issuecomment-1464944357. If `$a == $b` then `assert eq $a $b` should pass. Similarly if `$a != $b` then `assert ne $a $b` should pass. I do not agree that `assert ne 1 "2"` should fail.

I can update the tests to assert to the right behaviour or revert changing `assert ne`, then  it is up to you if you want implement the new style for it as well. (Just please document that it is intentionally not consistent with the `!=` operator.)

Comment (presidento) [reply crates/nu-utils/standard_library/std.nu:90]: I updated the tests.

File: crates/nu-utils/standard_library/std.nu
Comment (presidento) [reply crates/nu-utils/standard_library/std.nu]: As you wish.

**Comment by amtoine** [comment]
not sure what to do with these failing tests :thinking: 

i think `assert eq` and `assert ne` should raise an error when the types of the left and right operands are different but more importantly, with or without this type check, they shoudld definitely be consistent :yum: 

> **Note**
> maybe we need more feedback on this `assert ne`
> should `assert ne` not check the type because things like `1 != "2"` work?
> in that case, the comparison looks ill defined to me because `1` and `"2"` do not even have the same type, so yeah it is `true` that they are not equal, but to me they look not even comparable :open_mouth:

**Comment by bobhy** [comment]
It appears that Nushell will let you compare items of different types without complaining about type mismatch.
Therefore, I don't think `assert eq` or `assert ne` should insist that both sides be of the same type.
```nushell
〉1 == 1.0
true
〉true == 1
false
```
The `1 == 1.0` case isn't a crystal clear example of this, but surely `true` is boolean and `1` is not.

**Comment by bobhy** [comment]
Due to a race condition, this PR drops out the unit tests for `dirs` that were added earlier today.  Please rebase this on the very latest `main` to pick them up.

This is developer-hostile design, we shouldn't have such race conditions for unrelated sub-modules.  We need a more sophisticated test harness for stdlib so `dirs.nu` could simply provide a `test_dirs.nu` and not have to modify `test.nu`.

**Comment by presidento** [comment]
Who is the leader here? As I started to receive contradicting requests, I
need to know whom to follow.

1) May I rebase this branch? If yes, can I squash it into one commit?

2) Can I keep `assert $a != $b` consistent with `assert ne $a $b`?

**Comment by amtoine** [comment]
@fdncred @kubouch 

i think we need some help here to decide what to do with `assert eq` and `assert ne` :open_mouth: 

**TL;DR**: should `assert eq 1 "1"` and `assert ne 1 "2"` give errors or something else?

**Comment by amtoine** [comment]
@presidento 

i can say for sure @bobhy and i are not members of the core team :+1: 

the standard library is a very recent project with a lot of design being defined right now :open_mouth: 
i hope did not induce too much confusion about who was "in charge" and what had to be done here...

**Comment by presidento** [comment]
There is no problem. Sometimes we need to escalate and involve somebody in charge.

@fdncred @kubouch Please include `assert eq 1 1.0` to this topic as well. (At current implementation it gives error, while `1 == 1.0` is true.)

**Comment by kubouch** [comment]
So the question is whether the asserts should be type-sensitive, or not? I'm leaning towards mirroring the Nushell's `==`, `!=`, etc., but I don't have a strong preference.

The main point of this PR is to add message to assert, though, right? I would concentrate on that and we can leave the type sensitivity issue to further discussion / another PR. So, I think the PR shouldn't change the way it works currently, just add the custom message.

**Comment by amtoine** [comment]
> So the question is whether the asserts should be type-sensitive, or not? I'm leaning towards mirroring the Nushell's `==`, `!=`, etc., but I don't have a strong preference.
> 
> The main point of this PR is to add message to assert, though, right? I would concentrate on that and we can leave the type sensitivity issue to further discussion / another PR. So, I think the PR shouldn't change the way it works currently, just add the custom message.
>
agree with this :+1:

**Comment by presidento** [comment]
As you wish. Will do it on Monday.

**Comment by fdncred** [comment]
I'd defer to @kubouch's opinion.

**Comment by presidento** [comment]
(I reverted the change of behaviour and created additional issues for the mentioned topics. Now this PR is just about the detailed error messages and be able to add custom message to assert methods.)

**Review by amtoine — APPROVED** [review]

**Review by amtoine — CHANGES_REQUESTED** [review]
i wanted to approve because it looked great, but on further inspection, it does not behave as before i think :open_mouth:

@fdncred @kubouch 
i do not want to slow down this PR too much, should i investigate to make sure the behaviour is the same? :yum:

**Review by amtoine — APPROVED** [review]

**Comment by fdncred** [comment]
By reading the posts here, it seems like it should work before except `Now this PR is just about the detailed error messages and be able to add custom message to assert methods`. If that's not the case, then someone should investigate to try and understand what changed.

**Comment by amtoine** [comment]
> assert
> 
:+1: 

> assert eq
>
:+1: 

> # assert ne
> Previously `if types are equal: error` then `if not ($left != $right): error`.
> Now `if $left == $right: error` then `if $type_left != $type_right: error`.
>
this is where i disagree :yum: 

previously we had
```
assert ne [l, r] {
    assert (l.type == r.type)  # error when `l.type != r.type`
    assert (l.val != r.val)
}
```
but now we have
```
assert ne [l, r] {
    if (l.val == r.val): error
    if (l.type != r.type): error
}
```
which does not look the same to me :open_mouth: 

- maybe the definitions of `==` and `!=` in `nushell` do not show the non equivalence
- but as the type check is not after the value check, the two are clearly not equivalent

i hope i'm not becoming blind to what is written here, i've seen so much `assert`s these days :laughing:

**Comment by amtoine** [comment]
i'm sorry being that picky here...
i'm just really confused about the logical inversion introduced by this PR :thinking: 

before we had the `_assert` wrapper that, at least to me, made the checks in `assert`, `assert eq` and `assert ne` quite clear.
now all the boolean tests have been inverted and even some of the orders between the tests themselves, to accomodate to the new `_assertion-error`.

**Comment by presidento** [comment]
At the end of the day it is your code. If you don't like the style, feel free to "be inspired" with this PR, reject this one and resend a new one with the style you want. I will 👍 it. From functional point of view I am sure they are identical.

**Comment by amtoine** [comment]
> At the end of the day it is your code.
>
nope it is absolutely not, i know "my" asserts had a lot of flaws that need to be addressed :+1: 

> From functional point of view I am sure they are identical.
>
i did not check all the workings of `==` and `!=` in `nushell`, so i trust you on this :relieved: 

> If you don't like the style, feel free to "be inspired" with this PR, reject this one and resend a new one with the style you want. I will +1 it.
>
what i'm saying is that the whole logic inversion introduced here is confusing to me, merely to add better message that just require passing the `span` of the condition being tested :thinking: 
i might be the only one confused, but still, i was :wink: 

in the end, "changing the style" might have been better kept in a separate and clearer PR, leaving the "better messages" (and they truely are, don't get me wrong :muscle:) for this one.

@fdncred 
i think we're good :wink:

**Comment by presidento** [comment]
For having better error messages, the error-making part must be extracted, since it is the (only) common part. It has an inverse behaviour compared to `_assert`. So keeping the code as it was while extracting the inverse behaviour, well, can be done, but it would made the code more complex. I rewrote the `assert ne` command to use exactly the same checks as the original one had. I hope it makes clear that there were no behaviour changes here.

File: crates/nu-utils/standard_library/std.nu
Comment (fdncred) [inline crates/nu-utils/standard_library/std.nu:128]: Maybe check for inequality here so you don't have to have a no-op block after if?
```suggestion
    if (($left | describe) != ($right | describe)) {
```

File: crates/nu-utils/standard_library/std.nu
Comment (fdncred) [inline crates/nu-utils/standard_library/std.nu:133]: Same idea here. 

```suggestion
    if ($left == $right) {
```

**Comment by fdncred** [comment]
Other than not having no-op blocks, this code seems pretty good to me.

**Comment by presidento** [comment]
Done

**Comment by sholderbach** [comment]
Let's land this then :)

**Comment by amtoine** [comment]
:partying_face: noishe

**Pull request merged**
